{# #######################
  content.html
##########################

This block prints the main content of the report - it should loop through
the output from each module and print it in sections.

#}

{% set description_by_section = {
  "samtools-stats-alignment":
    "<p>Mapped vs. unmapped reads from <code>samtools stats</code>.
    <p>For a set of samples that have
    come from the same multiplexed library we should expect similar numbers of reads for each sample.
    Large differences in numbers might indicate issues during the library preparation process. Whilst
    large differences in read numbers may be controlled for in downstream processings (e.g. read count
    normalisation), you may wish to consider whether the read depths achieved have fallen below
    recommended levels depending on the applications (e.g. ~ 10 million reads for differential gene
    expression analysis.) If disambiguation was used these rates are only reflective of
    the component that was assigned to the primary species (typically human) and not the whole sequencing
    depth.</p>
    <p>A high-quality library should have ~95% of reads align. Low alignment rates could
    indicate contamination of samples (e.g. adapter sequences), low sequencing quality or other artefacts.
    These can be further investigated in the sequence level QC (e.g. from FastQC). If disambiguation was
    used, these percentages reflect fully aligned read pairs assigned to the primary species.
    if the percentage is not 100%, the lack is reflective of read pairs where one mate didn't align.
    </p>"
} %}

{% macro print_section(module_anchor, section) -%}
  <div class="mqc-section mqc-section-{{ module_anchor }}" id="{{ section.anchor }}">
  {% if section.content %}
    {% if section.anchor in description_by_section %}
      {{ description_by_section[section.anchor] }}
      {{ section.content|replace("<p>", "<p style='display: none;'>") }}
    {% else %}
      {{ section.content }}
    {% endif %}
  {% endif %}
  </div>
{%- endmacro %}

{% macro find_sections(group) %}
  {% for m in report.modules_output %}
    {% for s in m.sections %}
      {% if s.anchor == group.table %}
        {{ print_section(m.anchor, s) }}
      {% endif %}
    {% endfor %}
  {% endfor %}
  {% for anchor in group.anchors %}
    {% for m in report.modules_output %}
      {% if m.anchor == anchor %}
        <div id="mqc-module-section-{{ m.anchor }}" class="mqc-module-section">
        {{ m.intro if m.intro }}
      {% endif %}
      {% for s in m.sections %}
        {% if m.anchor == anchor or s.anchor == anchor %}
          <h3 id="{{ s.anchor }}">{{ s.name }}</h3>
          {{ print_section(m.anchor, s) }}
          <br>
        {% endif %}
      {% endfor %}
      {% if m.anchor == anchor %}
        </div>
      {% endif %}
    {% endfor %}
  {% endfor %}
{% endmacro %}

{% from "groups.html" import postalign_group, groups %}
{% for group in groups %}
  {% set sections = find_sections(group) %}
  {% if sections|trim %}
    <div id="mqc-module-section-{{ group.id }}" class="mqc-module-section">
    {% if group.id != postalign_group.id %}
      <h2 id="{{ group.id }}">{{ group.title }}</h2>
    {% endif %}
    {{ sections }}
    </div>
  {% endif %}
{% endfor %}

{#
  TODO: 5. add TargQC explanations from SOP
        6. more reasonable heatmaps (count outliers, or use knowledge of good coverage (80%?)
        8. Add more RNAseq tools?

Possible values in first table:
- Phenotype - combine with Math, see below
- Match - combine with Phenotype (see below) and move to the variants table)
Sample Name  Match
syn-tumor    syn-normal
syn-normal   [normal]


{#{% for m in report.modules_output %}#}
{#<div id="mqc-module-section-{{ m.anchor }}" class="mqc-module-section">#}
{#  <h2 id="{{ m.anchor }}">{{ m.name }}</h2>#}
{#  {{ m.intro if m.intro }}#}
{#  {% for s in m.sections %}#}
{#    <div class="mqc-section mqc-section-{{ m.anchor }}" id="{{ s.anchor }}">#}
{#      <h3 id="{{ s['anchor'] }}">{{ s['name'] }}</h3>#}
{#      {{ s['content'] if s['content'] }}#}
{#      {{ '<hr>' if not loop.last }}#}
{#    </div>#}
{#  {% endfor %}#}
{#</div>#}
{#{{ '<hr>' if not loop.last }}#}
{#{% endfor %}#}

{#
*Alignment Statistics*
 Target: <path>.bed
<table>
bcbio-fraction-coverage
qualimap-coverage-histogram
qualimap-cumulative-genome-fraction-coverage
qualimap-insert-size-histogram
qualimap-gc-distribution
samtools-...

*Variant Calling Statistics*
<table>
bcbio-fraction-variants
bcbio-fraction-qsignature
snpeff-...
#}